clang/include/clang/AST/ExprCXX.h

```diff
@@ -4902,6 +4902,188 @@ class BuiltinBitCastExpr final
   }
 };
 
+/// Represents a __builtin_num_fields expression.
+class SYCLBuiltinNumFieldsExpr : public Expr {
+  friend class ASTStmtReader;
+
+  SourceLocation Loc;
+  QualType SourceTy;
+
+public:
+  SYCLBuiltinNumFieldsExpr(SourceLocation Loc, QualType SourceTy,
+                           QualType RetTy)
+      : Expr(SYCLBuiltinNumFieldsExprClass, RetTy, VK_RValue, OK_Ordinary),
+        Loc(Loc), SourceTy(SourceTy) {
+    setDependence(computeDependence(this));
+  }
+
+  explicit SYCLBuiltinNumFieldsExpr(EmptyShell Empty)
+      : Expr(SYCLBuiltinNumFieldsExprClass, Empty) {}
+
+  QualType getSourceType() const { return SourceTy; }
+  unsigned getNumFields() const {
+    assert(!isTypeDependent() && !SourceTy->isDependentType());
+    assert(SourceTy->isRecordType());
+    const auto *RD = SourceTy->getAsRecordDecl();
+    assert(RD);
+    return static_cast<unsigned>(
+        std::distance(RD->field_begin(), RD->field_end()));
+  }
+
+  SourceLocation getBeginLoc() const { return getLocation(); }
```

Comment:
You need `getBeginLoc()` and `getEndLoc()` (compile errors from .inc-included code, otherwise). I can have it return `Loc` easily enough (though the `return getLocation();` is used by several other AST nodes and is equivalent).

Reply:
Ah, I see!  Interestingly I don't think I ran into that.